source/common/router/upstream_request.cc

```diff
@@ -450,8 +450,15 @@ void UpstreamRequest::onPoolReady(
     upstream_info.setUpstreamNumStreams(info.upstreamInfo().value().get().upstreamNumStreams());
   }
 
-  upstream_info.setUpstreamFilterState(std::make_shared<StreamInfo::FilterStateImpl>(
-      info.filterState().parent()->parent(), StreamInfo::FilterState::LifeSpan::Request));
+  // Upstream filters might have already created/set a filter state.
+  const StreamInfo::FilterStateSharedPtr& filter_state =
+      const_cast<StreamInfo::StreamInfo&>(info).filterState();
```

Comment:
const_cast seems to be "gotcha" here as it violates the `onPoolReady` contract. I see the problem though. Would it work to remove the constness from the `info` parameter? It will cause a much bigger diff, but at least it will not require this hack.

Reply:
Good point. I was not sure which is more disruptive - changing the interface or making the `const_cast` local change. I can try changing the interface and see how deep the change would be. Thanks for reviewing 💯

Reply:
@yanavlasov I have changed the interface but had to make more changes than I would have liked to. PTAL.

Also, the CI does not look happy. I am not sure if its related to the PR.